src/os/bluestore/BlueStore.h

```diff
@@ -467,50 +420,57 @@ class BlueStore : public ObjectStore,
     void _clear(BufferCacheShard* cache);
 
     // return value is the highest cache_private of a trimmed buffer, or 0.
-    int discard(BufferCacheShard* cache, uint32_t offset, uint32_t length) {
+    int discard(BufferCacheShard* cache,
+                uint32_t offset, uint32_t length) {
       std::lock_guard l(cache->lock);
       int ret = _discard(cache, offset, length);
       cache->_trim();
       return ret;
     }
-    int _discard(BufferCacheShard* cache, uint32_t offset, uint32_t length);
+    int _discard(BufferCacheShard* cache,
+                 uint32_t offset, uint32_t length);
 
-    void write(BufferCacheShard* cache, uint64_t seq, uint32_t offset, ceph::buffer::list&& bl,
+    void write(BufferCacheShard* cache,
+               uint64_t seq, uint32_t offset, ceph::buffer::list&& bl,
 	       unsigned flags) {
       std::lock_guard l(cache->lock);
       uint16_t cache_private = _discard(cache, offset, bl.length());
-      _add_buffer(cache, this, Buffer(this, Buffer::STATE_WRITING, seq, offset, std::move(bl),
-			     flags), cache_private, (flags & Buffer::FLAG_NOCACHE) ? 0 : 1, nullptr);
+      _add_buffer(cache,
+                  new Buffer(this, Buffer::STATE_WRITING, seq, offset, std::move(bl), flags),
+                  cache_private, (flags & Buffer::FLAG_NOCACHE) ? 0 : 1, nullptr);
```

Comment:
Why bring back heap allocated buffers here?

Reply:
That's because I switched the stuff to boost::intrusive.
Boost intrusive containers deal with heap allocated buffers since they do not perform object copying/allocation on insertion.

Reply:
Move ctor helps to avoid expensive data copying but stl containers need to allocate space to move the data to (from stack instance to heap one) anyway. So in boost you just do explicitly things (allocate and construct a collection entry) which stl does for you internally. In non-instrusive case you create a buffer on stack then insert it into a map/list. This makes stl to allocates a collection's entry which merges buffer and additional fields to keep that entry in the container (aka hooks in boost). Then it moves your buffer's content to internal entry's one . Thanks to move semantic it happens more efficiently but that's the only thing which movement improves. 
There is no performance degradation in boost insertion over stl one at all.